text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,368 @@
+    let buf = mem::take(&mut this.buf);
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    // SAFETY: double-drops are prevented by putting `this` in a ManuallyDrop that is never dropped
+    let inner = unsafe { ptr::read(&this.inner) };
+
+    (inner, buf)
+}
+```
+
+Now with `destructure!`:
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    destructure!(let Self { buf, inner, panicked } = self);
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    (inner, buf)
+}
+```
+
+We don't even need unsafe anymore :tada:
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A built-in macro with the following "signature" is added to the standard library:
+
+```rust
+#[macro_export]
+macro_rules! destructure {
+    ((let)? $pat:pat = $e:expr) => { /* built-in */ }
+}
+```
+
+The result of the expansion checks the following requirements and produces a compilation error if any of them are not met:
+- `$pat` must be an irrefutable pattern
+- `$e` must be an owned expression that can be matched by pattern `$pat`
+- `$e`'s type must be an ADT (`struct` or `enum`)
+- The type of `$e` must be constructable in the current context, i.e.
+    - All fields must be visible to the current module wrt privacy
+    - If the type is marked with `#[non_exhaustive]`, it must be defined in the current crate
+
+The semantic of this macro is equivalent to the following:
+1. Wrapping `$e` in a `ManuallyDrop`
+2. Copying **all** fields of `$e` (for enums all fields of the active variant)
+3. If the `let` token is present in the macro invocation, fields that are listed in `$pat` are binded to their respective bindings or patterns. Otherwise, they are assigned similarly to destructuring assignment
+4. If `..` is present in `$pat`, all fields that are not mentioned in it are dropped in the order of definition (simialrly to how [normal `let`/destructuring assignment works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3abd3aebd3378690ff3d2006e12d4120))
+---
+
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+When implementing this change as a macro expanding to a statement, we do not think there are many drawbacks. However, in the coming section with alternatives we discuss another possible expansion to a pattern which may have a few advantages, but also has many drawbacks which we discuss there.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
What about the really obvious alternative: "allow a plain `let Struct { .. } = foo;` without a macro". What's the rationale for preventing this in the first place? Specifically, we could allow this in exactly the cases where you propose the macro to work, i.e. when the type is constructible in the current context.

Reply:
I think it is important to syntactically mark the fact that `destructure!()` is a different operation than this. We talk about this in the guide level explanation. For types implementing Copy, `destructure!()` moves anyway and doesn't run the original value's destructor at all

Reply:
Hah apologies, I skipped over the guide-level explanation because I asumed the technical details would be in the reference-level section.

Here's what I was missing, which imo should be highlighted in the motivation section: `let Struct { <something> } = foo` does not reliably prevent `drop` from running, because only a non-`Copy` binding would force a move out of `foo`. Hence I seem to understand that the problem statement of this RFC includes "find a reliable way to disable drop".

It seems like allowing `let Struct { <something> } = foo` is pretty orthogonal in fact: we could allow it and this RFC would still make sense. The argument against is that users might think this disables `Drop` when it doesn't always. I would really like to see that spelled out in the "Alternatives" section.

Reply:
fair enough, I'll talk to @WaffleLapkin and we'll see what modifications to make. I'll let you know when we have so you can take another look if you'd like :)

Reply:
> Here's what I was missing, which imo should be highlighted in the motivation section: `let Struct { <something> } = foo` does not reliably prevent drop from running,

I do not understand the use of "reliably" here. This **never** disables `foo`'s drop because it's one of these situations:
- `Struct` implements `Drop` and `<something>` causes a move, which is currently error
- `<something>` doesn't cause a move, so drops are not affected at all
- `<something>` causes a move, but `Struct` does not implement `Drop`, which means that only the drop for the moved field is disabled (other fields still dropped)...

> Hence I seem to understand that the problem statement of this RFC includes "find a reliable way to disable drop".

Not really? The problem statement of this RFC is very specifically to allow moving fields out of `Drop` types. Disabling drop is a requirement for that -- you *have* to disable the drop, because it gets access to the fields and so can't work when a field is moved out.

The rest of the RFC is "how do we make this operation not confusing and least footguny". Simply allowing normal destructuring of types which implement `Drop` is a footgun, since the syntax doesn't clearly differentiate between disabling drop or copying fields out.

We might want to spell this out more explicitly though.

Reply:
> Simply allowing normal destructuring of types which implement Drop is a footgun, since the syntax doesn't clearly differentiate between disabling drop or copying fields out.

Yes, that's precisely the point I was trying to make! By "not reliable", I mean "allowing the obvious thing is footgunny because nothing would syntactically indicate whether you're moving out or not". Hence why I would like to see this spelled out, because that was not obvious to me on first read of the RFC.

Reply:
That could be remedied by just turning the hard error into a lint. Then you `#[expect]` the lint wherever you want the destructuring to happen and subsequently the drop getting avoided

Reply:
That could be remedied by just turning the hard error into a lint. Then you `#[expect]` the lint wherever you want the destructuring to happen and subsequently the drop getting avoided. Though that has the downside of adding a drop impl not causing existing destructuring sites to error in other crates